### ISSUE TITLE
Cart block - Third party plugin compatibility - Need replacement for woocommerce_cart_item_thumbnail hook

### ISSUE THREAD

**Issue by marcusig** [woocommerce/woocommerce-blocks#10527]
## Describe the bug

In the cart and checkout blocks, there is no way to replace the product image

## To reproduce

1. Go to https://demos.mklacroix.com/custom-laptop/
2. Select some options
3. Add to cart
4. Visit the cart comparison page (https://demos.mklacroix.com/cart-block-vs-cart-shortcode/)

or 

1. Use the filter `woocommerce_cart_item_thumbnail` to change the image of the cart items
2. e.g. 
```php
add_filter( 'woocommerce_cart_item_thumbnail', function() {
    return '<img src="https://user-images.githubusercontent.com/11990914/257510384-b6e0eff2-bcef-49d5-aa76-3c25c82f08ba.png">';
} );

```

## Expected behavior

The product image in the cart should be changed. Ideally there would be a PHP hook in the store API allowing to change the image. A JS hook could also be used, but it would not be ideal in my opinion, as it would require duplicating the logic which already exists in PHP.

## Screenshots

Cart Shortcode: | Cart Block:
--|--
![image](https://github.com/woocommerce/woocommerce-blocks/assets/11990914/b6e0eff2-bcef-49d5-aa76-3c25c82f08ba) | ![image](https://github.com/woocommerce/woocommerce-blocks/assets/11990914/b3674667-78f7-48e3-8db8-d6b06abc0077)

-   WordPress version: 6.3
-   WooCommerce version: 8.0.1
-   WooCommerce Blocks version: 10.8.2
-   Site language: EN
-   Any other plugins installed: [Product configurator for WooCommerce](https://wordpress.org/plugins/product-configurator-for-woocommerce/) and add-ons. 

## Additional context

As the developper of the plugins mentioned above, this issue is part of a series of issues brought forward in an effort to make them compatible with the cart and checkout blocks.

**Comment by marcusig** [comment]
@nielslange of course. Take for example this [demo product](https://demos.mklacroix.com/shop/custom-chair/), where the user can select various options in order to get the desired. 
The product image as set in the product object will be different from the selection the user makes. 
In [my plugin](https://wordpress.org/plugins/product-configurator-for-woocommerce/), an image is then generated to reflect the users selection. 
This selection should then be displayed in the cart / checkout, if the shop owner decides to. 

This is currently done using the hook `woocommerce_cart_item_thumbnail`, which has no effect in the blocks.

**Comment by andreagrillo** [comment]
Hi @nielslange, thanks for your reply. We have the same issues in a couple of our plugins. I think having that possibility before would be useful to be able to do that now, I thinking about backward compatibility.

**Comment by nielslange** [comment]
@marcusig Thanks for explaining your specific use case. This is very helpful. 

@andreagrillo Thanks for confirming that you're having the same issue in a couple of your plugins. Is your use case identical, and would you mind listing your plugins?

For now, I'll move this issue to our backlog.

**Comment by marcusig** [comment]
Thank you for looking into this.

**Comment by andreagrillo** [comment]
@nielslange Thanks! For sure! 

We're working to test all plugins. We have already tested this ones: 

YITH WooCommerce Product Add-ons (free version available on [here](https://wordpress.org/plugins/yith-woocommerce-product-add-ons/))

YITH WooCommerce Gift Card (free version available [here](https://wordpress.org/plugins/yith-woocommerce-gift-cards/))

YITH WooCommerce Product Bundles (free version available [here](https://wordpress.org/plugins/yith-woocommerce-product-bundles/))

**Comment by nielslange** [comment]
Thanks for sharing the affected extensions, @andreagrillo.

**Comment by marcusig** [comment]
@nielslange Any news on this? Is it likely that this will be sorted before the November release where the blocks are the "default experience"?

**Comment by nielslange** [comment]
> @nielslange Any news on this? Is it likely that this will be sorted before the November release where the blocks are the "default experience"?

Thanks for reaching out, @marcusig. This issue is still in our backlog and the release, that will land in the WooCommerce core to make the blocks the default experience, will be released early next week. I'm afraid, we won't be able to get this issue into that release.

That said, in https://github.com/woocommerce/woocommerce-blocks/issues/9302#issuecomment-1531186530 @opr shared the following code snippet to change the product image for a certain product name:

```php
add_filter( 'woocommerce_product_get_image_id', function( $id, $product ) {
	if ( $product->get_name() === 'Ticket attached to Event' ) {
		return '628'; // The ID of the event's featured image.
	}
	return $id;
}, 10, 2 );
```

I understood that in your use case the product image will be generated based on the buyer's selection. In this case, you'd have to store the generated image in the media library, fetch the media ID and the add it to the code snippet. I hope this helps to address your current use case.

**Comment by marcusig** [comment]
Thanks for the feedback @nielslange.
I'll take a look at that. The product itself doesn't hold any information about the customer's selection, it is in the cart item.

So in order for this to work, I suppose I would need to store the new image ID in the product meta, but it should obviously only affect the object in the cart. 
Is there a "right" way to do that?

**Comment by nielslange** [comment]
> So in order for this to work, I suppose I would need to store the new image ID in the product meta, but it should obviously only affect the object in the cart.

Where (or even if) to store the new product image ID depends on your business logic. Let's say you sell t-shirts and a buyer can add their name on the shirt. In this case, you might not want to store the the new product image ID in the product meta, as every order would add a new product image ID. In this example, I'd rather store the product image ID in the order meta table.

As for showing the product image on certain pages, let's assume that the product image should be shown both on the cart and checkout page. In this case, you'd use the following code snippet:

```php
add_filter( 'woocommerce_product_get_image_id', function( $id, $product ) {
	if ( ( is_cart() || is_checkout() ) && $product->get_name() === 'Ticket attached to Event' ) {
		return '628'; // The ID of the event's featured image.
	}
	return $id;
}, 10, 2 );
```

For better readability, you could also use this code snippet, which is identical to the one before, with the exception that it's different formatted:
```php
/**
 * Retrieves the modified or original image ID for a WooCommerce product.
 *
 * @param int        $id The current image ID.
 * @param WC_Product $product The WooCommerce product object.
 * @return int The modified or original image ID.
 */
function custom_product_image_id( $id, $product ) {
	// Check if on cart or checkout pages.
	$is_relevant_page = is_cart() || is_checkout();

	// Check if the product name matches the specific product.
	$is_target_product = $product->get_name() === 'Ticket attached to Event';

	// If on relevant pages and for the target product, return the custom ID.
	if ( $is_relevant_page && $is_target_product ) {
		return '628';
	}

	return $id;
}
add_filter( 'woocommerce_product_get_image_id', 'custom_product_image_id', 10, 2 );
```
